### PR TITLE
Wrapper around Net::DNS which enforces timeouts

### PR DESCRIPTION
**Pull request by marcbradshaw**

### CODE REVIEW

**Comment by marcbradshaw** [comment]
some test failures around critic to be fixed.

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package Mail::Milter::Authentication::Resolver;
+use strict;
+use warnings;
+# VERSION
+use base 'Net::DNS::Resolver';
+
+=head1 DESCRIPTION
+
+Superclass for Net::DNS::Resolver, Versions of Net::DNS::Resolver from 1.03 up (to at least
```
Comment (robn) [inline lib/Mail/Milter/Authentication/Resolver.pm]: You mean "subclass"?

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  my $handler = $self->{_handler};
+  my $config = $handler->config();
+  my $timeout = $config->{'dns_timeout'};
+
+  my $return;
+
+  eval {
+    $handler->set_handler_alarm( ( $timeout + 0.1 ) * 1000000 ); # 0.1 seconds over that passed to Net::DNS::Resolver
+    if ( $what eq 'query' ) {
+      $return = $self->SUPER::send( @_ );
```
Comment (robn) [inline lib/Mail/Milter/Authentication/Resolver.pm]: Think you mean `::query`?

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    elsif ( $what eq 'search' ) {
+      $return = $self->SUPER::send( @_ );
```
Comment (robn) [inline lib/Mail/Milter/Authentication/Resolver.pm]: And `::search` I guess.

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    elsif ( $what eq 'send' ) {
+      $return = $self->SUPER::send( @_ );
+    }
+    $handler->reset_alarm();
+  };
+
```
Comment (robn) [inline lib/Mail/Milter/Authentication/Resolver.pm]: Tiny window here for `SIGALRM` to be delivered outside of the eval, before the alarm is reset? Probably not a ton we can do about it though?

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    $handler->set_handler_alarm( ( $timeout + 0.1 ) * 1000000 ); # 0.1 seconds over that passed to Net::DNS::Resolver
+    if ( $what eq 'query' ) {
+      $return = $self->SUPER::send( @_ );
+    }
+    elsif ( $what eq 'search' ) {
+      $return = $self->SUPER::send( @_ );
+    }
+    elsif ( $what eq 'send' ) {
+      $return = $self->SUPER::send( @_ );
+    }
```
Comment (robn) [inline lib/Mail/Milter/Authentication/Resolver.pm]: You could instead do `$return = $self->SUPER::$what(@_)`. Bit iffy maybe, but saves repeating the list of methods.

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  my $self = shift;
+  my $what = shift;
+  my $handler = $self->{_handler};
+  my $config = $handler->config();
+  my $timeout = $config->{'dns_timeout'};
+
+  my $return;
+
+  eval {
+    $handler->set_handler_alarm( ( $timeout + 0.1 ) * 1000000 ); # 0.1 seconds over that passed to Net::DNS::Resolver
```
Comment (mmcclimon) [inline lib/Mail/Milter/Authentication/Resolver.pm]: It is slightly confusing to me that `$config->{dns_timeout}` is in seconds but this is in microseconds. (The comment helps, though.)

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      $return = $self->SUPER::send( @_ );
+    }
+    $handler->reset_alarm();
+  };
+
+  if ( my $error = $@ ) {
+    $handler->reset_alarm();
+    my $type = $handler->is_exception_type( $error );
+    if ( $type ) {
+      if ( $type eq 'Timeout' ) {
```
Comment (mmcclimon) [inline lib/Mail/Milter/Authentication/Resolver.pm]: (nit) I would probably rather read this as `if ($type && $type eq 'Timeout')`

File: lib/Mail/Milter/Authentication/Resolver.pm [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+to Net::DNS::Resolver
+
+=cut
+
+{
+  sub new {
+    my $class = shift;
+    my %args = @_;
+    my $self = $class->SUPER::new( @_ );
+    $self->{ _handler } = $args{_handler};
```
[the remark this answers is not in the document]
Comment (marcbradshaw) [reply lib/Mail/Milter/Authentication/Resolver.pm]: It shouldn't matter, but it won't hurt to do it anyway

File: lib/Mail/Milter/Authentication/Resolver.pm (hunk 6 above)
Comment (marcbradshaw) [reply lib/Mail/Milter/Authentication/Resolver.pm]: Mostly legacy from when sub second timeouts didn't matter

File: lib/Mail/Milter/Authentication/Resolver.pm (hunk 4 above)
Comment (marcbradshaw) [reply lib/Mail/Milter/Authentication/Resolver.pm]: I don't think there is potential for them to go uncaught this way.

File: lib/Mail/Milter/Authentication/Resolver.pm (hunk 5 above)
Comment (marcbradshaw) [reply lib/Mail/Milter/Authentication/Resolver.pm]: Changed to be more succinct, but without the iffy

**Pull request merged**